smdebug/tensorflow/keras.py

```diff
@@ -366,6 +385,90 @@ def _add_metric(self, metric_name, metric_value: tf.Tensor = None):
             coll.set_tensor_ref(TensorRef.from_non_graph_var(metric_name))
         self.tensor_to_collections[metric_name] = {coll}
 
+    def save_custom_tensor(self, tensor_name, tensor_value, collections_to_write):
+        for collection in collections_to_write:
+            self.custom_tensors_to_save[tensor_name] = (tensor_value, collection)
+
+    def _save_custom_tensors_post_step(self):
+        for tensor_name in self.custom_tensors_to_save:
+            tensor_value, collection_names = self.custom_tensors_to_save[tensor_name]
+            self._save_tensor(tensor_name, tensor_value, collection_names)
+
+    def _save_tensor(self, tensor_name, tensor_value, collections_to_write):
+        if isinstance(collections_to_write, set) is False:
+            collections_to_write = {collections_to_write}
+        collections_to_save = self._get_collections_to_save_for_step()
+        for collection in collections_to_save:
+            if match_inc(tensor_name, collection.include_regex):
+                collections_to_write.add(collection)
+        self._initialize_writers(only_initialize_if_missing=True)
+        tensor_refs = []
+        if isinstance(tensor_value, values.PerReplica):
+            for t in tensor_value._values:
+                tensor_ref = TensorRef.from_non_graph_var(tensor_name)
+                tensor_refs.append((tensor_ref, t))
+        else:
+            tensor_ref = TensorRef.from_non_graph_var(tensor_name)
+            tensor_refs.append((tensor_ref, tensor_value))
+
+        for tensor_ref, t in tensor_refs:
+            for collection in collections_to_write:
+                if isinstance(collection, str):
+                    collection = self.get_collection(collection)
+                collection.set_tensor_ref(tensor_ref)
+            self._save_for_tensor(tensor_name, t, check_before_write=True)
+
+    def save_smdebug_logs(self, logs):
+        if logs is None:
+            return
+
+        model_input_tensor_id = 0
+
+        for key in logs:
+            if "smdebug_" in key:
+                if key in ModelOutputs:
+                    tensors_to_save = []
+                    export_name = get_model_output_export_name(key)
+                    tensors_to_save.append((export_name, logs[key]))
+                    collections_to_write = (
+                        {self.get_collection(CollectionKeys.OUTPUTS)}
+                        if self._is_collection_being_saved_for_step(CollectionKeys.OUTPUTS)
+                        else set()
+                    )
+                    for t_name, t_value in tensors_to_save:
+                        self._save_tensor(t_name, t_value, collections_to_write)
```

Comment:
these 2 lines are in every if else condition except SMDEBUG_LAYER_OUTPUTS_KEY. can this be made common and be called just before function ends?

Reply:
Done